source/common/watchdog/abort_action/abort_action.cc

```diff
@@ -0,0 +1,53 @@
+#include "common/watchdog/abort_action/abort_action.h"
+
+#include "envoy/thread/thread.h"
+
+#include "common/common/assert.h"
+#include "common/common/fmt.h"
+#include "common/common/logger.h"
+#include "common/protobuf/utility.h"
+#include "common/thread/terminate_thread.h"
+
+namespace Envoy {
+namespace Watchdog {
+namespace AbortAction {
+
+AbortAction::AbortAction(envoy::watchdog::abort_action::v3alpha::AbortActionConfig& config,
+                         Server::Configuration::GuardDogActionFactoryContext& /*context*/)
+    : config_(config){};
+
+void AbortAction::run(
+    envoy::config::bootstrap::v3::Watchdog::WatchdogAction::WatchdogEvent /*event*/,
+    const std::vector<std::pair<Thread::ThreadId, MonotonicTime>>& thread_last_checkin_pairs,
+    MonotonicTime /*now*/) {
+
+  if (thread_last_checkin_pairs.empty()) {
+    ENVOY_LOG_MISC(warn, "Watchdog AbortAction called without any thread.");
+    return;
+  }
+
+  // The following lines of code won't be considered covered by code coverage
+  // tools since they would run in DEATH tests.
+  const auto& thread_id = thread_last_checkin_pairs[0].first;
+  const std::string tid_string = thread_id.debugString();
+  ENVOY_LOG_MISC(error, "Watchdog AbortAction terminating thread with tid {}.", tid_string);
+
+  if (Thread::terminateThread(thread_id)) {
+    // Successfully signaled to thread to terminate, sleep for wait_duration.
+    absl::SleepFor(absl::Milliseconds(PROTOBUF_GET_MS_OR_DEFAULT(config_, wait_duration, 0)));
+  } else {
+    // Failed to send the signal, abort?
+    ENVOY_LOG_MISC(error, "Failed to terminate tid {}", tid_string);
+  }
+
+  // Abort from the action since the signaled thread hasn't yet crashed the process.
+  // Panicing in the action gives flexibility since it doesn't depend on
+  // external code to kill the process if the signal fails.
+  PANIC(fmt::format(
```

Comment:
It may make sense to install these default actions in the guarddog itself instead of doing so by modifying config earlier in the process.  Doing so would allow us to get more consistent behavior in smaller tests and avoid potentially showing the config modifications when someone accesses the proxy config via the admin handler.

Reply:
That was a great suggestion -- having this logic captured within just the guarddog simplifies a lot of the design.